### ISSUE TITLE
Optimize trace generation

### ISSUE THREAD

**Issue by Al-Kindi-0** [0xPolygonMiden/miden-vm#1558]
Currently, trace generation, both main and auxiliary, takes roughly about 10% of the total proving time. We benchmark and profile this step in order to get an idea of bottlenecks.
Witness generation is an inherently sequential process but there are still strategies that one could explore in order to improve things. One such approach is multi-stage building of the main trace through several passes, where for example the first pass could fill the minimal amount of trace values needed in order to be able to fill the hasher chiplet in parallel.

**Comment by plafer** [comment]
Additional context: https://github.com/0xPolygonMiden/miden-vm/pull/1533#issuecomment-2449052138

**Comment by bobbinth** [comment]
I think a good target here would be to get main trace generation to run at at least 50 MHz on something like an M1. And a stretch goal could be 100 MHz. I think it should be double but would require quite a bit of investigation and potential refactoring.

**Comment by bobbinth** [comment]
One of the approach to investigate here is to completely split execution and trace generation. So, we'd have 2 components:

1. Executor: execute a program from start to finish as quickly as possible while maintaining minimal required state.
2. Trace generator: build the execution trace for a contiguous set of VM cycles.

The main idea is that the executor will run in a single thread while multiple trace generators could run concurrently in multiple threads:

![image](https://github.com/user-attachments/assets/b48a6e85-5fb8-4e28-a73c-76d1df78ccb4)

The key question is which parts of the state trace generators would need to generate the trace. Here is my initial stab at this:

1. Stack state - this can be tracked pretty efficiently and can probably be cloned with minimal overhead when the executor is done with a given segment.
2. Memory state - this would also need to be cloned but probably would be somewhat more expensive then cloning the stack state. One potential optimization here is to clone only the part of the memory which is "touched" during the current segment.
3. Advice provider state - cloning this one may require quite a bit of work as advice provider may contain quite a bit of data. However, the Merkle Store component is "append only" and so does not need to be cloned. If we can make Advice Map append-only as well, we can avoid cloning here too. Cloning of the advice stack should be pretty cheap.
4. Decoder state - I haven't thought this through completely, but I think the state here should be relatively small and shouldn't be too expensive to clone.

Assuming trace generation is at least 10x more expensive than execution, we should be able to get close to 5x-10x speed up on high-end laptops, and even more on sever-grade machines (i.e., the more cores we can throw at it, the faster we'll generate the trace).

**Comment by plafer** [comment]
Note: removed from 0.14.0 since this issue includes more than just the initial version of the fast processor.

**Comment by plafer** [comment]
I looked a bit more into this. Basically `rayon` is more amenable to this kind of data-parallel computation than `tokio`'s `Stream`s - mainly because with `tokio` I am not able to pass in a non-static `&mut buffer` slice to a thread (see https://github.com/tokio-rs/tokio/issues/3162#issuecomment-745218533).

The "downside" is that rayon's parallel iterator expects the whole `Vec` that we iterate on to be present (i.e. would be a vector of "checkpoints" in our case), so that implies that the fast processor would need to be done running the program before we can start trace generation. But assuming a 100MHz fast processor, we will execute a $2^{16}$ trace in 0.6ms, and a $2^{18}$ trace in 2ms - clearly not the bottleneck, so this should be fine.

However @bobbinth we had discussed using this as an opportunity to make the processor `async`. Given the above, I now feel like the processor would still be fundamentally synchronous - and adding an `async` API would then best be dealt with in a separate issue.

**Comment by bobbinth** [comment]
> mainly because with `tokio` I am not able to pass in a non-static `&mut buffer` slice to a thread

Were you thinking that each task here would just work on a different part of the same buffer? If so, this may not be possible because for some parts of the trace (e.g., range checker, chiplets), we can't construct the trace until the very end.

My original thinking was that each task builds its own segment of the trace, and then once the last segment is built, we assemble everything into one trace (this would mostly involve memory copies). The assumption here is that doing memcopies at the end is going to be faster than trying to run all trace segment generators at the end.

So, maybe there are actually 3 components here:

- Processor which executes the program as fast as possible and creates checkpoints.
- The processor sends the checkpoints to a trace generator (which would run in a separate thread).
- The trace generator would then send of checkpoints to individual tasks (each would be run in a separate thread), and as these tasks return trace segments, it would keep track of them, and put them into a single trace after execution is complete.

Basically, trace generator would be responsible for coordinating the work of trace segment generators and building the "global" part of the trace.

**Comment by bobbinth** [comment]
> This would make sense if the fast processor's time to execute the whole program is a non-negligible percentage of the time it takes to generate the trace. But as I argued above, napkin math suggests that it would be a trivial percentage, and hence waiting for the fast processor to generate all checkpoints before starting trace generation wouldn't affect the overall performance (with the benefit that we'd be able then to use `rayon`).

I'd like to test this on a sample transaction. The main reason is that executing a transaction involves a bunch of hashing and some Merkle path verification, and so, executing transactions would probably run at a much slower pace than 100 MHz (e.g., if it takes 5 ms to execute a transaction that would be awesome, but it may end up bing more than that).

**Comment by plafer** [comment]
Looking more into this, I agree that a `tokio`-based approach where we don't have to wait for the fast processor to be done before starting trace generation is favorable since it's more flexible. That is, current or future workloads where the fast processor is not so fast (e.g. lots of hashing) would still run "optimally" with the tokio-based approach, whereas they would be much slower with the `rayon`-based approach.

This opens up the can of worms that is sync vs async in Rust. Although we have our `winter-maybe-async` crate that partially solves this problem, there are drawbacks with that approach (well explained in [this article](https://nullderef.com/blog/rust-async-sync/)). The main question I have is: **if we move the processor to be async, do we still need to expose a sync API**?

Depending on the answer to that question, we have a few options:
1. Only expose an `async` API
2. use `winter-maybe-async` to expose both a sync and async API
    - Although the main drawback is that the `is-async` feature is not "additive", as explored in the aforementioned article
3. use the [`async-generic`](https://crates.io/crates/async-generic) crate
    - no feature flags, and millions of downloads (which I recall was a concern when choosing not to use [`maybe-async`](https://crates.io/crates/maybe-async))
4. expose a `blocking` module, similar to [how `reqwest` does it](https://docs.rs/reqwest/latest/reqwest/blocking/index.html)

Needless to say, the simplest approach would be to only expose an `async` API. So before I really start thinking hard about the best way to support both sync and async, it'd be great to know that there's a real use case for still supporting a `sync` API.

cc @bobbinth, possibly @bitwalker @PhilippGackstatter @phklive

**Comment by bitwalker** [comment]
I'm curious about the choice of async for this versus plain multi-threading + channels. Async does not mix well with compute-heavy tasks, it is a paradigm really best suited for I/O-bound tasks. It sounds to me like the only real reason we're considering the async route is so that when trace generation begins, it can be suspended when further work by the processor task must be done to make progress. One can achieve the same goal much more simply with threads and channels, with the additional benefit that you can then freely use task-parallel (e.g. rayon) or dataflow techniques to parallelize the computations themselves as you see fit.

I would bet cold hard cash that making the processor async will slow it down by an order of magnitude.

---

Setting that aside, I would personally go the route of exposing a sync wrapper around async APIs, but skip making the async part a Cargo feature - the underlying implementation is async, so there is no benefit to making it a Cargo feature when you don't get the benefit of avoiding the async dependencies. I think that is acceptable with Miden crates, whereas with a much more general purpose library, I'd want to do the extra work to allow eliding the async stuff if a user so chooses. Doing so is a ton more work, but worth it in that circumstance, since it makes the library much more useful in different contexts. Miden crates don't need to worry about that IMO.

**Comment by plafer** [comment]
> I'm curious about the choice of async for this versus plain multi-threading + channels.

The main reason we were thinking of async is that it has been requested for the processor to have an async API, and so we saw this rework as an opportunity to kill two birds with one stone. But indeed in my dummy project where I'm trying out ideas, I ended up basically using tokio tasks + channels, where the same could be achieved with just channels + threads.

> I would bet cold hard cash that making the processor async will slow it down by an order of magnitude.

If this is the case, then this is an argument in favor of being "sync-first", and possibly e.g. expose silly async wrappers full of `block_on()`'s.

**Comment by bitwalker** [comment]
> The main reason we were thinking of async is that it has been requested for the processor to have an async API,

Makes sense, that request was for browser-based execution I'm assuming? If that's the case, we'd really need to make the processor _actually_ async, otherwise there isn't any tangible benefit (i.e. the primary need for it in browser-based programs is because if you don't yield to the event loop, or do so frequently enough, your whole UI freezes up until it does). 

An alternative solution for that use case, is to use service workers, i.e. basically a background thread - the downside is that it is more complicated to start and interact with code running in a worker, compared to the simplicity of just calling a JS function and waiting for it to return.

and so we saw this rework as an opportunity to kill two birds with one stone. But indeed in my dummy project where I'm trying out ideas, I ended up basically using tokio tasks + channels, where the same could be achieved with just channels + threads.

> If this is the case, then this is an argument in favor of being "sync-first", and possibly e.g. expose silly async wrappers full of `block_on()`'s.

I don't really see any purpose in ever defining such functions in our crates - leave it up to the downstream crate to define such functions based on the async runtime they are using. I suppose the only exception would be if the implementation of such a function needs access to internal details, but that feels like a sign that either it shouldn't be public, or something is private that should be public). IMO it only ever makes sense to do one of the following:

1. Only sync (simplest, least flexible)
2. Only async (most flexible, but not convenient at all from sync contexts)
3. Async implementation, with an API that provides sync alternatives where sensible (addresses the ergonomics of 2)
4. User chooses between 1 and 2 based on Cargo feature (most complicated, most useful)

**Comment by bobbinth** [comment]
> To summarize, it's because the host can then make async calls to some data store that is itself async and does not offer a sync option (which is the case for the standard browser DB impl iiuc).

Yes, this is the main reason. Basically, currently we cannot support things that require event handlers to be async (which is something we need in WASM). A good example of this is Falcon signature generation. Right now, the way the wallet guys solve this is by loading private keys into process memory so that requests can be made sync. This is not a safe approach and won't work long term. So, we have 3 options:

1. Make execution async (proving can still be sync).
2. Drop support for transaction execution in WASM (i.e., always rely on delegated proving).
3. Figure out if the approach proposed by @Mirko-von-Leipzig works.

What to do here really depends on how impactful/complex each option is:

- If option 3 works and is not too difficult to implement, we could go with that.
- If option 1 doesn't result in a noticeable performance drop, we could go with that too.

So, I guess one of the main questions to answer is whether bringing in the async runtime would actually have a negative impact on proving time. And her again, we really only need async for execution (and maybe trace generation) - proving could stay as it is.

A separate question is how this could work in Plonky3 context. That is, can trace generation be async once we move to Plonky3 (and can we even run proof generation in WASM with Plonky3). cc @adr1anh and @Al-Kindi-0

**Comment by Al-Kindi-0** [comment]
> A separate question is how this could work in Plonky3 context. That is, can trace generation be async once we move to Plonky3 (and can we even run proof generation in WASM with Plonky3). cc @adr1anh and @Al-Kindi-0

The crates that make up Plonky 3 are all `no_std`, moreover, as we are going to need to implement our own prover, using the building blocks provided by Plonky3, we will have the freedom to choose how we go about addressing the questions raised in this issue. This in particular includes the question of trace generation, which we will for sure be able to make it work in WASM.

**Comment by PhilippGackstatter** [comment]
> It sounds to me like you were exploring using shared memory between the main thread and the worker thread

Yes, although not directly. The async runtimes I was trying to use needed it to work, but I forgot too many details to describe what exactly I was trying to do. I'm also not very versed in Web APIs so I might have used it incorrectly. So I agree it would be good to explore this again.

**Comment by bobbinth** [comment]
Interesting! It seems like going with `tokio`-based approach makes sense. One question: out of 4.4ms, how big is the final step of copying all fragments into one big matrix?

**Comment by plafer** [comment]
> One question: out of 4.4ms, how big is the final step of copying all fragments into one big matrix?

About 0.5ms. 

But just so that everyone is on the same page: we should consider this 0.5ms solely as an absolute quantity rather than in comparison with the 4.4ms. That is, the fact that the entire trace generation in the benchmark takes 4.4ms is meaningless, as it speaks more to my choice of generating a new random number for each trace cell. For example, if I instead fill it with 0's instead, the pre-copy time drops to 0.9ms (from 3.9ms).

**Comment by Mirko-von-Leipzig** [comment]
Async implies the call doesn't block the thread for any serious amount of time. And that length of time isn't defined anywhere, but I recall seeing something like a an upper bound of 100us be suggested. As in, execution shouldn't exceed 100us before hitting another `.await` point.

I would be _quite_ surprised if proving or anything of this nature is what anyone would realistically consider "async". In the context of a webpage for example, this would block a webpage unless its offloaded to a separate thread.

**Comment by plafer** [comment]
> Async implies the call doesn't block the thread for any serious amount of time. And that length of time isn't defined anywhere, but I recall seeing something like an upper bound of 100us be suggested. As in, execution shouldn't exceed 100us before hitting another .await point.

Ah right good point, although [Claude](https://claude.ai/) suggests that this upper bound is closer to ~1-10ms (which would make sense for e.g. a single-threaded web application trying to draw at 60 Hz, or every 16 ms). In any case, we would probably go over that 10ms in workloads with a lot of hashing, and would be safer to use [tokio's `spawn_blocking`](https://docs.rs/tokio/latest/tokio/index.html#cpu-bound-tasks-and-blocking-code). Although I'll need to go back to the benchmarks to see how that affects performance.

Thank you for pointing that out!

EDIT: but then as you mentioned, there's no actual `async` work, which is an argument for just making this API "sync-first".

**Comment by Mirko-von-Leipzig** [comment]
My suggestion would be to remove IO from the implementation (sans-IO). Otherwise you will always end up with async/sync concerns since you have multiple conflicting IO requirements depending on the use case.

If there is no IO required, then it should stay be sync.

> Ah right good point, although [Claude](https://claude.ai/) suggests that this upper bound is closer to ~1-10ms (which would make sense for e.g. a single-threaded web application trying to draw at 60 Hz, or every 16 ms).

I imagine Claude et al would be _very_ web biased which makes sense. The reason to minimize blocking is so that other tasks/threads don't get starved. If we have multiple trace's being generated concurrently, or if this "fills" all the cores, then everything else halts.
